drivers/pwm/pwm_nrfx.c

```diff
@@ -12,86 +12,245 @@ LOG_MODULE_REGISTER(pwm_nrfx);
+	 * or NRF_PWM_CHANNEL_COUNT if not initialized through DTS.
+	 */
+	channel = pwm_channel_map(pdata, pwm);
+	if (channel == NRF_PWM_CHANNEL_COUNT) {
+		LOG_ERR("PWM pin %d not enabled through DTS configuration.",
+			pwm);
 		return -EINVAL;
 	}
 
-	if (period_cycles != pdata->top_value) {
-		if (period_cycles > PWM_COUNTERTOP_COUNTERTOP_Msk) {
-			return -EINVAL;
-		}
-
-		pdata->top_value = period_cycles;
-		nrf_pwm_configure(pconfig->pwm.p_registers,
-				  pconfig->config.base_clock,
-				  pconfig->config.count_mode,
-				  pdata->top_value);
+	/* Check if period_cycle is either matching currently used, or
+	 * possible to use with our prescaler options.
+	 */
+	ret = pwm_period_check_and_set(pdata, pwm_instance, pwm, period_cycles);
+	if (ret) {
+		LOG_ERR("Incompatible period %d, already in use: %d",
+			period_cycles, pdata->period_cycles);
+		return ret;
 	}
 
-	if (pulse_cycles > pdata->top_value) {
+	/* Check if pulse is bigger than period, fail if so */
+	if (pulse_cycles > period_cycles) {
 		return -EINVAL;
 	}
 
-	/* Modify only the COMPARE bits while preserving the POLARITY
-	 * bit that controls the inversion of the channel.
-	 * For more details see product specification.
+	/* Store new pulse value bit[14:0], and polarity bit[15] for channel. */
+	pdata->current[channel] = (
+		(pdata->current[channel] & NRFX_PWM_POLARITY_MSK)
+		| (pulse_cycles / pdata->prescaler_divisor)
+		);
+
+	/* If Channel is off/fully on (duty 0% or 100%), also set GPIO register
+	 * since this will the setting if we in the future disable the
+	 * peripheral when no channels are active.
 	 */
-	pdata->current[pwm] = ((pdata->current[pwm]
-				& ~PWM_COUNTERTOP_COUNTERTOP_Msk)
-				| pulse_cycles);
+	if (pulse_cycles == 0 || pulse_cycles == period_cycles) {
+		/* If pulse 0% and pin not inverted, set LOW.
+		 * If pulse 100% and pin inverted, set LOW.
+		 * If pulse 0% and pin inverted, set HIGH.
+		 * If pulse 100% and pin not inverted, set HIGH.
+		 */
+		bool channel_inverted_state =
+			get_pin_inversion_state(
+				pdata->config.output_pins[channel]);
+
+		bool pulse_0_and_not_inverted =
+			(pulse_cycles == 0)
+			&& !channel_inverted_state;
+		bool pulse_100_and_inverted =
+			(pulse_cycles == period_cycles)
+			&& channel_inverted_state;
+
+		if (pulse_0_and_not_inverted || pulse_100_and_inverted) {
+			NRF_GPIO->OUTCLR = BIT(pwm);
+		} else {
+			NRF_GPIO->OUTSET = BIT(pwm);
+		}
+	}
+
+	/* Check if all channels are off (duty 0% or 100%) */
+	if (!pwm_channel_active(pdata)) {
+		/* If peripheral in use and all channels off: disable. */
+		if (pdata->pwm_instance_in_use) {
+			nrf_pwm_disable(pwm_instance);
+			pdata->pwm_instance_in_use = false;
+		}
+	} else {
+		/* A PWM Channel is active:
+		 * Enable peripheral if disabled, and start sequence.
+		 */
+		if (!pdata->pwm_instance_in_use) {
+			nrf_pwm_enable(pwm_instance);
+			pdata->pwm_instance_in_use = true;
+		}
+
+		/* Start sequence */
+		nrfx_pwm_simple_playback(&pconfig->pwm,
```

Comment:
This should be inside the `if` block above.

Reply:
This has to be called anytime we have a new sequence to play. So it needs to be outside the if block.

If pwm is not in use, start it. Then start new playback.
If pwm is in use, we simply have to start a new playback to get the new values.

I agree that this is confusing, i will add some comments to make it more clear what is being done.

Reply:
Right, thanks. Now i understood what you mean. The nrfx layer is using both sequences and connecting them through the use of shorts, it will keep loading in new values all the time from the same array and thus we only have to start the sequence when its not already running. Will fix it.